### ISSUE TITLE
DDPOptimizer split graphs should pass symbolic shape variables as arguments

### ISSUE THREAD

**Issue by sujoysaraswati** [pytorch/pytorch#116956]
### 🐛 Describe the bug

The below test case uses DDPOptimizer with dynamic shapes. It splits the forward graph such that one of the split graphs has an input tensor with a symbolic expression in a dim. The split FX graph doesn't pass the symbolic shape variables in the arguments. For direct symbolic shapes, inductor is able to decipher the variable from the input tensor dims. However, for the case below, inductor fails when the dim is a symbolic expression.

Shouldn't FX graphs declare every variable, even if they represent symbolic shapes, explicitly as input argument or as an IR node output, before its use?

```
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.nn import Linear
from torch.optim import SGD
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.nn.parallel import DistributedDataParallel as DDP
from torch._dynamo.backends.distributed import DDPOptimizer
import torch._dynamo as dynamo
from torch._inductor.compile_fx import compile_fx

os.environ["TORCH_COMPILE_DEBUG"] = "0"
dynamo.reset()

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'
    dist.init_process_group(backend="GLOO", rank=rank, world_size=world_size)

def cleanup():
    dist.destroy_process_group()

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.lin1 = Linear(3 * 1024, 2 * 1024, bias=False)
        self.lin2 = Linear(2 * 1024, 2 * 1024, bias=False)
        self.lin3 = Linear(2 * 1024, 3 * 1024, bias=False)

    def forward(self, x):
        t1 = self.lin1(x).relu().add(1.0)
        t2 = self.lin2(t1).relu().pow(2.0)
        c1 = torch.cat((t1, t2))
        t3 = self.lin3(c1).relu().mul(2.0)
        return t3

def simple(rank, world_size, iterations):
    setup(rank, world_size)
    model = ToyModel()
    model = DDP(model, bucket_cap_mb=70)

    intensor = torch.rand(8 * 1024, 3 * 1024)
    ddp_optimizer = DDPOptimizer(
              bucket_bytes_cap=int(100 * 1024 * 1024),
              backend_compile_fn=compile_fx
          )

    @torch._dynamo.optimize(ddp_optimizer.compile_fn, dynamic=True)
    def opt_fn(inp):
          return model(inp)

    optim = SGD(model.parameters(), lr=0.1)

    for i in range(iterations):
          out = opt_fn(intensor)
          out.float().sum().backward()
          optim.step()
          optim.zero_grad()

    print(f"Done for rank :: {rank}")
    cleanup()

if __name__ == "__main__":
    WORLD_SIZE = 2
    mp.start_processes(simple,
        args=(WORLD_SIZE, 5),
        nprocs=WORLD_SIZE,
        join=True,
        start_method="fork")
``` 

This fails with the following error -

```
[rank0]:[2024-01-08 05:55:10,075] [0/0] torch._dynamo.variables.torch: [WARNING] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
[rank1]:[2024-01-08 05:55:10,146] [0/0] torch._dynamo.variables.torch: [WARNING] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
/usr/local/lib/python3.10/dist-packages/torch/overrides.py:110: UserWarning: 'has_cuda' is deprecated, please use 'torch.backends.cuda.is_built()'
  torch.has_cuda,
/usr/local/lib/python3.10/dist-packages/torch/overrides.py:111: UserWarning: 'has_cudnn' is deprecated, please use 'torch.backends.cudnn.is_available()'
  torch.has_cudnn,
/usr/local/lib/python3.10/dist-packages/torch/overrides.py:117: UserWarning: 'has_mps' is deprecated, please use 'torch.backends.mps.is_built()'
  torch.has_mps,
/usr/local/lib/python3.10/dist-packages/torch/overrides.py:118: UserWarning: 'has_mkldnn' is deprecated, please use 'torch.backends.mkldnn.is_available()'
  torch.has_mkldnn,
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
---------------------------------------------------------------------------
ProcessRaisedException                    Traceback (most recent call last)
[<ipython-input-17-50ec36e203b3>](https://localhost:8080/#) in <cell line: 65>()
     65 if __name__ == "__main__":
     66     WORLD_SIZE = 2
---> 67     mp.start_processes(simple,
     68         args=(WORLD_SIZE, 5),
     69         nprocs=WORLD_SIZE,

1 frames
[/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py](https://localhost:8080/#) in join(self, timeout)
    161         msg = "\n\n-- Process %d terminated with the following error:\n" % error_index
    162         msg += original_trace
--> 163         raise ProcessRaisedException(msg, error_index, failed_process.pid)
    164 
    165 

ProcessRaisedException: 

-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py", line 74, in _wrap
    fn(i, *args)
  File "<ipython-input-17-50ec36e203b3>", line 57, in simple
    out = opt_fn(intensor)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 328, in _fn
    return fn(*args, **kwargs)
  File "<ipython-input-17-50ec36e203b3>", line 52, in opt_fn
    return model(inp)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/parallel/distributed.py", line 1519, in forward
    else self._run_ddp_forward(*inputs, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/parallel/distributed.py", line 1355, in _run_ddp_forward
    return self.module(*inputs, **kwargs)  # type: ignore[index]
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "<ipython-input-17-50ec36e203b3>", line 32, in forward
    def forward(self, x):
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 328, in _fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 678, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 284, in __call__
    raise e
  File "/usr/local/lib/python3.10/dist-packages/torch/fx/graph_module.py", line 274, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "<eval_with_key>.204", line 7, in forward
    submod_1 = self.compiled_submod_1(submod_0);  submod_0 = None
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/backends/distributed.py", line 335, in forward
    x = self.submod(*args)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 328, in _fn
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 3905, in forward
    return compiled_fn(full_args)
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 1482, in g
    return f(*args)
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 2527, in runtime_wrapper
    all_outs = call_func_with_args(
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 1506, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 1482, in g
    return f(*args)
  File "/usr/local/lib/python3.10/dist-packages/torch/autograd/function.py", line 539, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 3010, in forward
    fw_outs = call_func_with_args(
  File "/usr/local/lib/python3.10/dist-packages/torch/_functorch/aot_autograd.py", line 1506, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/usr/local/lib/python3.10/dist-packages/torch/_inductor/codecache.py", line 374, in __call__
    return self.get_current_callable()(inputs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_inductor/codecache.py", line 401, in _run_from_cache
    return compiled_graph.compiled_artifact(inputs)
  File "/tmp/torchinductor_root/ep/cepmz3ofpsywz5ucqoluwu37y7tozhk4bgh7yhucxdw5w3ulnn6c.py", line 54, in call
    assert_size_stride(primals_2, (2*s0, 2048), (2048, 1))
NameError: name 's0' is not defined
```

The DDPOptimizer splits the graph where the last linear is in a separate graph -
```
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] DDPOptimizer produced the following bucket assignments:
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] ┌─────────┬────────────┬───────────────────────┐
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] │   Index │   Size (b) │ Param Names           │
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] ├─────────┼────────────┼───────────────────────┤
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] │       0 │   25165824 │ L__self___lin3_weight │
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] ├─────────┼────────────┼───────────────────────┤
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] │       1 │   41943040 │ L__self___lin2_weight │
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] ├─────────┼────────────┼───────────────────────┤
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] │         │            │ L__self___lin1_weight │
[rank1]:[2024-01-08 05:41:55,714] [1/0] torch._dynamo.backends.distributed: [DEBUG] └─────────┴────────────┴───────────────────────┘
```
The split submodule graph for this, where the input to the graph is the cat output that has a shape (2*s0, 3072) -
```
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] ---submod_1 graph---
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG] graph():
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %cat : [num_users=1] = placeholder[target=cat]
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %l__self___lin3 : [num_users=1] = call_module[target=L__self___lin3](args = (%cat,), kwargs = {})
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %relu : [num_users=1] = call_method[target=relu](args = (%l__self___lin3,), kwargs = {})
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     %mul : [num_users=1] = call_method[target=mul](args = (%relu, 2.0), kwargs = {})
[rank1]:[2024-01-08 05:41:55,725] [1/0] torch._dynamo.backends.distributed.__ddp_graphs: [DEBUG]     return mul
```
When inductor generates the callable for this, it doesn't declare/find the s0 variable but uses it in the callable code, and the runtime exception comes due to this -

```
[rank1]:[2024-01-08 05:42:12,647] [1/0] torch._inductor.graph.__output_code: [DEBUG] def call(args):
[rank1]:[2024-01-08 05:42:12,647] [1/0] torch._inductor.graph.__output_code: [DEBUG]     primals_1, primals_2 = args
[rank1]:[2024-01-08 05:42:12,647] [1/0] torch._inductor.graph.__output_code: [DEBUG]     args.clear()
[rank1]:[2024-01-08 05:42:12,647] [1/0] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(primals_1, (3072, 2048), (2048, 1))
[rank1]:[2024-01-08 05:42:12,647] [1/0] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(primals_2, (2*s0, 2048), (2048, 1))
[rank1]:[2024-01-08 05:42:12,647] [1/0] torch._inductor.graph.__output_code: [DEBUG]     buf0 = empty_strided((2*s0, 3072), (3072, 1), device='cpu', dtype=torch.float32)
...
```

Note that when the input tensor has a symbolic shape (as the split submodule 0, which has an input tensor with shape s0, 2048)), inductor generates the code to extract the dim from the input argument, as seen in the callable generated from submodule 0 -

```
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG] def call(args):
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG]     primals_1, primals_2 = args
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG]     args.clear()
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG]     primals_2_size = primals_2.size()
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG]     s0 = primals_2_size[0]
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(primals_1, (2048, 2048), (2048, 1))
[rank1]:[2024-01-08 05:42:12,027] [1/0] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(primals_2, (s0, 2048), (2048, 1))
...
``` 

### Versions

PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.58+-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             2
On-line CPU(s) list:                0,1
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                         6
Model:                              79
Thread(s) per core:                 2
Core(s) per socket:                 1
Socket(s):                          1
Stepping:                           0
BogoMIPS:                           4399.99
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          32 KiB (1 instance)
L1i cache:                          32 KiB (1 instance)
L2 cache:                           256 KiB (1 instance)
L3 cache:                           55 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0,1
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable; SMT Host state unknown
Vulnerability Meltdown:             Vulnerable
Vulnerability Mmio stale data:      Vulnerable
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.1.0+cu121
[pip3] torchaudio==2.1.0+cu121
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.1.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @kadeng @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @msaroufim @bdhirsh @anijain2305 @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

**Comment by wconstab** [comment]
Thanks for the report @sujoy! 

can you confirm if you're using `config.optimize_ddp_lazy_compile` True or False (in dynamo config)?

**Comment by sujoysaraswati** [comment]
> optimize_ddp_lazy_compile

No, I used the default config and the PyTorch version I am using doesn't have this flag in the dynamo config.

**Comment by wconstab** [comment]
Hmm, then can you confirm what version you are using? Bc you might have the lazy compile on by default before we added the flag to disable it, or you might have the non lazy compile before we added the lazy compile..

**Comment by sujoysaraswati** [comment]
> PyTorch version: 2.1.0

Yes, it is PyTorch version: 2.1.0. 
```
print(f"torch version : {torch.__version__}")
```
shows -
```
torch version : 2.1.0+cu121
```
I checked that the dynamo config doesn't have the optimize_ddp_lazy_compile flag -

```
import torch._dynamo.config as config
print(f"is optimize_ddp_lazy_compile present? {hasattr(config, 'optimize_ddp_lazy_compile')}")
```
prints -
```
is optimize_ddp_lazy_compile present? False
```

**Comment by alanhe151220037** [comment]
@wconstab
I can repro this in a recent version `torch 2.3.0a0+git60f4114` and this error could be avoided when I enable `optimize_ddp_lazy_compile`.
I'm not sure whether this bug is associated with the previous problem fixed in #116292. But I do have some finding about why `s0` is missed in the submodule graph cache. In the codegen function here (line 758), `shape` is `2*s0` but only `s0` is in `needed`. The function here is not designed to deal with the situation that input `shape` is a formula rather than a single symbol which may occur in submodule.
https://github.com/pytorch/pytorch/blob/9d5954e2a942de684b069fcf8c50b2f3572129e9/torch/_inductor/codegen/wrapper.py#L754-L762

**Comment by wconstab** [comment]
if we enable lazy compile, we use real tensor outputs from first inductor subgraph to feed into compilation of second subgraph.  i'm wondering if we'd still have a shape formula on input there, or if it forces subgraph inputs to be fresh symints.  cc @jon-chuang @bdhirsh

**Comment by ezyang** [comment]
@sujoysaraswati has diagnosed the problem perfectly. All we need to do is make the splitter pass dynamic integers as arguments to all the split subgraphs as necessary. This is not too hard to do, just needs to be done.

**Comment by wconstab** [comment]
how do we do this? i can work on a PR if you give me a pointer

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/117406

**Comment by sujoysaraswati** [comment]
> #117406

Thanks @ezyang for the fix. Will this be available with PyTorch 2.2 release?

**Comment by ezyang** [comment]
No, it missed the branch cut :(
